File: sdks/python/tests/e2e/runner/echo_app.py
"""Agent app used by the runner e2e tests.

Contains two entrypoints:
- echo: basic echo function
- echo_config: echo with a configurable greeting (used in mask tests)
"""

import threading

import opik


@opik.track(entrypoint=True)
def echo(message: str) -> str:
    print(f"echo stdout: {message}")
    return f"echo: {message}"


class EchoConfig(opik.Config):
    greeting: str


@opik.track(entrypoint=True)
def echo_config(message: str) -> str:
    client = opik.Opik()
    cfg = client.get_or_create_config(
        fallback=EchoConfig(greeting="fallback-greeting"),
    )
    return f"{cfg.greeting}: {message}"


if __name__ == "__main__":
    # Analogous to uvicorn.run() in a FastAPI app: @opik.track(entrypoint=True) registers
    # handlers at import time, and this blocks the main thread so the runner loop stays
    # alive to process jobs.
    threading.Event().wait()

File: sdks/typescript/tests/integration/runner/echo_app.ts
/**
 * Agent app used by the runner integration tests.
 *
 * Contains two entrypoints:
 * - echo: basic echo function
 * - echo_config: echo with a configurable greeting (used in mask tests)
 *
 * Stays alive via setInterval to simulate a long-running server process.
 */

import { OpikClient as Opik } from "@/client/Client";
import { track } from "@/decorators/track";

const echo = track(
  { entrypoint: true, name: "echo", params: [{ name: "message", type: "string" }] },
  async (message: string): Promise<string> => {
    return `echo: ${message}`;
  }
);

const echo_config = track(
  { entrypoint: true, name: "echo_config", params: [{ name: "message", type: "string" }] },
  async (message: string): Promise<string> => {
    const client = new Opik();
    await client.createConfig({ greeting: "default-greeting" });
    const cfg = await client.getOrCreateConfig({
      fallback: { greeting: "fallback-greeting" },
    });
    return `${cfg.greeting}: ${message}`;
  }
);

// Keep the process alive until killed by the test harness (SIGTERM via afterAll).
// process.stdin.resume() holds the event loop open regardless of timer unref state.
process.stdin.resume();

process.on("unhandledRejection", (err) => {
  console.error("Unhandled rejection in echo_app:", err);
});

void echo;
void echo_config;
